harness/determined/profiler.py

```diff
@@ -0,0 +1,666 @@
+        if not self.is_enabled:
+            return
+
+        self.current_batch_idx = new_batch_idx
+        self.sys_metric_collector_thread.update_batch_idx(self.current_batch_idx)
+
+        # Check if we should start collecting metrics
+        if (
+            not self.is_active
+            and not self.has_finished
+            and self.current_batch_idx >= self.start_on_batch
+        ):
+            self._begin_collection()
+
+        # Check if we should stop collecting metrics due to batch idx being exceeded
+        if (
+            self.is_active
+            and self.end_after_batch is not None
+            and self.current_batch_idx > self.end_after_batch
+        ):
+            self._end_collection()
+
+    def _begin_collection(self) -> None:
+        if not self.is_enabled:
+            return
+
+        # Note: due to its simplicity, sender_thread doesn't need to be activated
+        self.sys_metric_collector_thread.activate()
+        # TODO: Activate TimingBatcher as well
+        self.shutdown_timer.activate()
+        self.has_started = True
+
+    def _end_collection(self) -> None:
+        """
+        Stop collecting data and shut down child threads. This function can be invoked due to the
+        max batch idx being exceeded, due to timeout or due to the ProfilingAgent shutting down as
+        the harness exits, so the function needs to be threadsafe and idempotent.
+        """
+
+        with self.shutdown_lock:
+            if self.has_finished:
+                return
+
+            if self.is_enabled:
+                # Shut down in reverse creation order
+                self.shutdown_timer.kill()
+                self.shutdown_timer.join()
+
+                if self.sysmetrics_is_enabled:
+                    self.sys_metric_collector_thread.kill()
+                    self.sys_metric_collector_thread.join()
+
+                # TODO: Shut down TimingBatcher as well
+
+                self.sender_thread.kill()
+                self.sender_thread.join()
+
+            self.has_finished = True
+
+    def record_timing(self, timing: float) -> None:
+        if not self.is_active:
+            return
+        # TODO: Add new timing to TimingBatcher
+
+
+class CustomTimer(threading.Thread):
+    """
+    Version of threading.Timer that can be cleaned up if the timer is no longer needed
+    ```
+    timer = CustomTimer(300, callback_fn, callback_args=[1,2,3])
+    timer.start()  # start the thread, not the timer
+    timer.begin_timer()
+
+    # After 300 seconds, callback_fn will be executed
+
+    # If we need to clean up the timer
+    timer.kill()  # This is idempotent and will work fine if the timer has gone off
+    ```
+    """
+
+    def __init__(
+        self, duration: int, callback: Callable, callback_args: Optional[List[Any]] = None
+    ):
+        self.duration = duration
+        self.callback = callback
+        self.callback_args = callback_args if callback_args is not None else []
```

Comment:
callback_args seems unnecessary; I think args should be bound into a closure.  `labmda: my_fn(my_args)` is pretty common in python.

Reply:
Fixed - I just removed the concept of args altogether since it's just dead code now